cpp/src/fil/cudf_to_matrix.cu

```diff
@@ -0,0 +1,151 @@
+    case cudf::type_id::DICTIONARY32:            ///< Dictionary type using int32 indices
+    case cudf::type_id::STRING:                  ///< String elements
+    case cudf::type_id::LIST:                    ///< List elements
+    case cudf::type_id::STRUCT:                  ///< Struct elements
+    case cudf::type_id::NUM_TYPE_IDS:            ///< Total number of type ids
+    default:
+      // bad code, return NAN for now
+      return NAN;
+  }
+}
+
+// cudf::column_view contains more info than we really need, and requires an extra pointer
+// dereference in each GPU thread
+struct device_col {
+  // contains scale for fixed-point columns
+  cudf::data_type dtype;
+  const void* value;
+  const uint8_t* valid_mask;  // bit N set to 1 if Nth value is valid (cudf convention)
+};
+
+// blocks are square tiles. If last column reached, waste the remainder of tile.
+#define TILE_SIZE std::size_t(32)
+
+// is lossy a scalar or vector (1 per column)?
+__global__ void cast_and_transpose(
+  float* row_major, const device_col* cols, std::size_t n_cols, std::size_t n_rows, int* lossy)
+{
+#define SHMEM_TILE_COLS (TILE_SIZE + std::size_t(1))
+  __shared__ float shmem_row_major[TILE_SIZE * SHMEM_TILE_COLS];
+  // thread index preamble
+  std::size_t tile_row0 = (std::size_t)TILE_SIZE * blockIdx.x;
+  std::size_t tile_col0 = (std::size_t)TILE_SIZE * blockIdx.y;
+
+  char row1_in_tile = threadIdx.x, col1_in_tile = threadIdx.y;
+  std::size_t row1 = tile_row0 + row1_in_tile;
+  std::size_t col1 = tile_col0 + col1_in_tile;
+  if (row1 < n_rows && col1 < n_cols) {
+    device_col column = cols[col1];
+    // load from gmem. check if valid and convert
+    float val = cast_scalar(column.value, row1, column.dtype, lossy);
+    if (column.valid_mask != nullptr && !ML::fil::fetch_bit(column.valid_mask, row1)) val = NAN;
+    // shmem write index will be the non-contiguous one
+    shmem_row_major[row1_in_tile * SHMEM_TILE_COLS + col1_in_tile] = val;
+  }
+  // move shmem -> gmem
+  __syncthreads();
+  // same tile, difference only within tile
+  char row2_in_tile = threadIdx.y, col2_in_tile = threadIdx.x;
+  std::size_t row2 = tile_row0 + row2_in_tile;
+  std::size_t col2 = tile_col0 + col2_in_tile;
+  if (row2 < n_rows && col2 < n_cols)
+    row_major[row2 * n_cols + col2] =
+      shmem_row_major[row2_in_tile * SHMEM_TILE_COLS + col2_in_tile];
+}
+
+// similar to interleave_columns()
+
+namespace ML {
+
+void cudf_to_row_major(const raft::handle_t& h,
+                       float* row_major,
+                       const std::vector<cudf::column_view>& cols)
+{
+  std::size_t n_cols = cols.size();
+  std::size_t n_rows = cols[0].size();
+
+  std::vector<device_col> h_cols;
+  for (const cudf::column_view& cv : cols) {
+    h_cols.push_back({cv.type(), cv.head(), reinterpret_cast<const uint8_t*>(cv.null_mask())});
+    ASSERT(static_cast<std::size_t>(cv.size()) == n_rows,
+           "internal error: creating matrix out of unequally sized columns");
```

Comment:
Are unequally-sized columns allowed in cuDF? If no, it's indeed an internal error. If yes, it's unsupported or incorrect input.